File: sdks/python/hatchet_sdk/features/webhooks.py
from __future__ import annotations

import asyncio
from typing import Any

from hatchet_sdk.clients.rest.api.webhook_api import WebhookApi
from hatchet_sdk.clients.rest.api_client import ApiClient
from hatchet_sdk.clients.rest.models.v1_create_webhook_request import (
    V1CreateWebhookRequest,
)
from hatchet_sdk.clients.rest.models.v1_create_webhook_request_api_key import (
    V1CreateWebhookRequestAPIKey,
)
from hatchet_sdk.clients.rest.models.v1_create_webhook_request_base import (
    V1CreateWebhookRequestBase,
)
from hatchet_sdk.clients.rest.models.v1_create_webhook_request_basic_auth import (
    V1CreateWebhookRequestBasicAuth,
)
from hatchet_sdk.clients.rest.models.v1_create_webhook_request_hmac import (
    V1CreateWebhookRequestHMAC,
)
from hatchet_sdk.clients.rest.models.v1_update_webhook_request import (
    V1UpdateWebhookRequest,
)
from hatchet_sdk.clients.rest.models.v1_webhook import V1Webhook
from hatchet_sdk.clients.rest.models.v1_webhook_api_key_auth import V1WebhookAPIKeyAuth
from hatchet_sdk.clients.rest.models.v1_webhook_auth_type import V1WebhookAuthType
from hatchet_sdk.clients.rest.models.v1_webhook_basic_auth import V1WebhookBasicAuth
from hatchet_sdk.clients.rest.models.v1_webhook_hmac_auth import V1WebhookHMACAuth
from hatchet_sdk.clients.rest.models.v1_webhook_list import V1WebhookList
from hatchet_sdk.clients.rest.models.v1_webhook_source_name import V1WebhookSourceName
from hatchet_sdk.clients.rest.tenacity_utils import tenacity_retry
from hatchet_sdk.clients.v1.api_client import BaseRestClient


class CreateWebhookRequest(V1CreateWebhookRequestBase):
    auth: V1WebhookBasicAuth | V1WebhookAPIKeyAuth | V1WebhookHMACAuth

    def _to_api_payload(self) -> V1CreateWebhookRequest:
        payload = self.model_dump(by_alias=True, exclude_none=True)
        payload["auth"] = self.auth.model_dump(by_alias=True)
        request_payload: (
            V1CreateWebhookRequestBasicAuth
            | V1CreateWebhookRequestAPIKey
            | V1CreateWebhookRequestHMAC
            | None
        ) = None
        if isinstance(self.auth, V1WebhookBasicAuth):
            payload["authType"] = V1WebhookAuthType.BASIC
            request_payload = V1CreateWebhookRequestBasicAuth.from_dict(payload)
        elif isinstance(self.auth, V1WebhookAPIKeyAuth):
            payload["authType"] = V1WebhookAuthType.API_KEY
            request_payload = V1CreateWebhookRequestAPIKey.from_dict(payload)
        else:
            payload["authType"] = V1WebhookAuthType.HMAC
            request_payload = V1CreateWebhookRequestHMAC.from_dict(payload)
        if request_payload is None:
            raise ValueError("failed to build create webhook request from payload")
        return V1CreateWebhookRequest(request_payload)


class WebhooksClient(BaseRestClient):
    """
    The webhooks client provides methods for managing incoming webhooks in Hatchet.

    Webhooks allow external systems to trigger Hatchet workflows by sending HTTP
    requests to dedicated endpoints. This enables real-time integration with
    third-party services like GitHub, Stripe, Slack, or any system that can send
    webhook events.
    """

    def _wa(self, client: ApiClient) -> WebhookApi:
        return WebhookApi(client)

    async def aio_list(
        self,
        limit: int | None = None,
        offset: int | None = None,
        webhook_names: list[str] | None = None,
        source_names: list[V1WebhookSourceName] | None = None,
    ) -> V1WebhookList:
        """
        List webhooks for a given tenant.

        :param limit: The maximum number of webhooks to return.
        :param offset: The number of webhooks to skip before starting to collect the result set.
        :param webhook_names: A list of webhook names to filter by.
        :param source_names: A list of source names to filter by.

        :return: A list of webhooks matching the specified criteria.
        """
        return await asyncio.to_thread(
            self.list, limit, offset, webhook_names, source_names
        )

    def list(
        self,
        limit: int | None = None,
        offset: int | None = None,
        webhook_names: list[str] | None = None,
        source_names: list[V1WebhookSourceName] | None = None,
    ) -> V1WebhookList:
        """
        List webhooks for a given tenant.

        :param limit: The maximum number of webhooks to return.
        :param offset: The number of webhooks to skip before starting to collect the result set.
        :param webhook_names: A list of webhook names to filter by.
        :param source_names: A list of source names to filter by.

        :return: A list of webhooks matching the specified criteria.
        """
        with self.client() as client:
            v1_webhook_list = tenacity_retry(
                self._wa(client).v1_webhook_list, self.client_config.tenacity
            )
            return v1_webhook_list(
                tenant=self.tenant_id,
                limit=limit,
                offset=offset,
                webhook_names=webhook_names,
                source_names=source_names,
            )

    def get(self, webhook_name: str) -> V1Webhook:
        """
        Get a webhook by its name.

        :param webhook_name: The name of the webhook to retrieve.

        :return: The webhook with the specified name.
        """
        with self.client() as client:
            v1_webhook_get = tenacity_retry(
                self._wa(client).v1_webhook_get, self.client_config.tenacity
            )
            return v1_webhook_get(
                tenant=self.tenant_id,
                v1_webhook=webhook_name,
            )

    async def aio_get(self, webhook_name: str) -> V1Webhook:
        """
        Get a webhook by its name.

        :param webhook_name: The name of the webhook to retrieve.

        :return: The webhook with the specified name.
        """
        return await asyncio.to_thread(self.get, webhook_name)

    def create(
        self,
        source_name: V1WebhookSourceName,
        name: str,
        event_key_expression: str,
        auth: V1WebhookBasicAuth | V1WebhookAPIKeyAuth | V1WebhookHMACAuth,
        scope_expression: str | None = None,
        static_payload: dict[str, Any] | None = None,
        return_event_as_response_payload: bool = True,
    ) -> V1Webhook:
        """
        Create a new webhook.

        :param source_name: The source name identifying the external system sending webhook events.
        :param name: The name of the webhook.
        :param event_key_expression: A CEL expression used to extract the event key from the incoming payload.
        :param auth: The authentication configuration for the webhook (basic auth, API key, or HMAC).
        :param scope_expression: An optional CEL expression used to extract the scope from the incoming payload.
        :param static_payload: An optional static payload to merge into every triggered event.
        :param return_event_as_response_payload: Whether to return the triggered event as the response payload of the webhook request.

        :return: The created webhook.
        """
        validated_payload = CreateWebhookRequest(
            sourceName=source_name,
            name=name,
            eventKeyExpression=event_key_expression,
            scopeExpression=scope_expression,
            staticPayload=static_payload,
            auth=auth,
            returnEventAsResponsePayload=return_event_as_response_payload,
        )
        with self.client() as client:
            return self._wa(client).v1_webhook_create(
                tenant=self.tenant_id,
                v1_create_webhook_request=validated_payload._to_api_payload(),
            )

    async def aio_create(
        self,
        source_name: V1WebhookSourceName,
        name: str,
        event_key_expression: str,
        auth: V1WebhookBasicAuth | V1WebhookAPIKeyAuth | V1WebhookHMACAuth,
        scope_expression: str | None = None,
        static_payload: dict[str, Any] | None = None,
        return_event_as_response_payload: bool = True,
    ) -> V1Webhook:
        """
        Create a new webhook.

        :param source_name: The source name identifying the external system sending webhook events.
        :param name: The name of the webhook.
        :param event_key_expression: A CEL expression used to extract the event key from the incoming payload.
        :param auth: The authentication configuration for the webhook (basic auth, API key, or HMAC).
        :param scope_expression: An optional CEL expression used to extract the scope from the incoming payload.
        :param static_payload: An optional static payload to merge into every triggered event.
        :param return_event_as_response_payload: Whether to return the triggered event as the response payload of the webhook request.

        :return: The created webhook.
        """
        return await asyncio.to_thread(
            self.create,
            source_name,
            name,
            event_key_expression,
            auth,
            scope_expression,
            static_payload,
            return_event_as_response_payload,
        )

    def update(
        self,
        webhook_name: str,
        event_key_expression: str | None = None,
        scope_expression: str | None = None,
        static_payload: dict[str, Any] | None = None,
        return_event_as_response_payload: bool | None = None,
    ) -> V1Webhook:
        """
        Update a webhook by its name.

        :param webhook_name: The name of the webhook to update.
        :param event_key_expression: An updated CEL expression used to extract the event key from the incoming payload.
        :param scope_expression: An updated CEL expression used to extract the scope from the incoming payload.
        :param static_payload: An updated static payload to merge into every triggered event.
        :param return_event_as_response_payload: Whether to return the triggered event as the response payload of the webhook request.

        :return: The updated webhook.
        """
        with self.client() as client:
            return self._wa(client).v1_webhook_update(
                tenant=self.tenant_id,
                v1_webhook=webhook_name,
                v1_update_webhook_request=V1UpdateWebhookRequest(
                    eventKeyExpression=event_key_expression,
                    scopeExpression=scope_expression,
                    staticPayload=static_payload,
                    returnEventAsResponsePayload=return_event_as_response_payload,
                ),
            )

    async def aio_update(
        self,
        webhook_name: str,
        event_key_expression: str | None = None,
        scope_expression: str | None = None,
        static_payload: dict[str, Any] | None = None,
        return_event_as_response_payload: bool | None = None,
    ) -> V1Webhook:
        """
        Update a webhook by its name.

        :param webhook_name: The name of the webhook to update.
        :param event_key_expression: An updated CEL expression used to extract the event key from the incoming payload.
        :param scope_expression: An updated CEL expression used to extract the scope from the incoming payload.
        :param static_payload: An updated static payload to merge into every triggered event.
        :param return_event_as_response_payload: Whether to return the triggered event as the response payload of the webhook request.

        :return: The updated webhook.
        """
        return await asyncio.to_thread(
            self.update,
            webhook_name,
            event_key_expression,
            scope_expression,
            static_payload,
            return_event_as_response_payload,
        )

    def delete(self, webhook_name: str) -> V1Webhook:
        """
        Delete a webhook by its name.

        :param webhook_name: The name of the webhook to delete.

        :return: The deleted webhook.
        """
        with self.client() as client:
            return self._wa(client).v1_webhook_delete(
                tenant=self.tenant_id,
                v1_webhook=webhook_name,
            )

    async def aio_delete(self, webhook_name: str) -> V1Webhook:
        """
        Delete a webhook by its name.

        :param webhook_name: The name of the webhook to delete.

        :return: The deleted webhook.
        """
        return await asyncio.to_thread(self.delete, webhook_name)

File: cmd/hatchet-cli/cli/webhooks.go
package cli

import (
	"fmt"
	"strings"

	tea "github.com/charmbracelet/bubbletea"
	"github.com/spf13/cobra"

	"github.com/hatchet-dev/hatchet/cmd/hatchet-cli/cli/internal/config/cli"
	"github.com/hatchet-dev/hatchet/cmd/hatchet-cli/cli/internal/styles"
	"github.com/hatchet-dev/hatchet/cmd/hatchet-cli/cli/tui"
	"github.com/hatchet-dev/hatchet/pkg/client/rest"
)

var webhookSourceNames = []rest.V1WebhookSourceName{
	rest.GENERIC,
	rest.GITHUB,
	rest.LINEAR,
	rest.SLACK,
	rest.STRIPE,
	rest.SVIX,
}

var webhookHMACAlgorithms = []rest.V1WebhookHMACAlgorithm{
	rest.MD5,
	rest.SHA1,
	rest.SHA256,
	rest.SHA512,
}

var webhookHMACEncodings = []rest.V1WebhookHMACEncoding{
	rest.BASE64,
	rest.BASE64URL,
	rest.HEX,
}

var webhooksCmd = &cobra.Command{
	Use:     "webhooks",
	Aliases: []string{"webhook"},
	Short:   "Manage webhooks",
	Long:    `Commands for listing, inspecting, creating, updating, and deleting webhooks.`,
	Run:     func(cmd *cobra.Command, args []string) { _ = cmd.Help() },
}

var webhooksListCmd = &cobra.Command{
	Use:   "list",
	Short: "List webhooks",
	Long:  `List webhooks. Without --output json, launches the interactive TUI. With --output json, outputs raw JSON.`,
	Example: `  # Launch interactive TUI (default)
  hatchet webhooks list --profile local

  # JSON output
  hatchet webhooks list -o json`,
	Run: func(cmd *cobra.Command, args []string) {
		isJSON := isJSONOutput(cmd)
		selectedProfile, hatchetClient := clientFromCmd(cmd)

		if !isJSON {
			tuiM := newTUIModel(selectedProfile, hatchetClient)
			tuiM.currentViewType = ViewTypeWebhooks
			tuiM.currentView = tui.NewWebhooksView(tuiM.ctx)
			p := tea.NewProgram(tuiM, tea.WithAltScreen())
			if _, err := p.Run(); err != nil {
				cli.Logger.Fatalf("error running TUI: %v", err)
			}
			return
		}

		ctx := cmd.Context()
		tenantUUID := clientTenantUUID(hatchetClient)
		limit := int64(50)
		offset := int64(0)
		resp, err := hatchetClient.API().V1WebhookListWithResponse(ctx, tenantUUID, &rest.V1WebhookListParams{
			Limit:  &limit,
			Offset: &offset,
		})
		if err != nil {
			cli.Logger.Fatalf("failed to list webhooks: %v", err)
		}
		if resp.JSON200 == nil {
			cli.Logger.Fatalf("unexpected response from API (status %d)", resp.StatusCode())
		}

		printJSON(resp.JSON200)
	},
}

var webhooksGetCmd = &cobra.Command{
	Use:   "get <webhook-name>",
	Short: "Get a webhook",
	Long:  `Get details about a single webhook by name. Outputs raw JSON.`,
	Args:  cobra.ExactArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		name := args[0]
		_, hatchetClient := clientFromCmd(cmd)
		tenantUUID := clientTenantUUID(hatchetClient)

		resp, err := hatchetClient.API().V1WebhookGetWithResponse(cmd.Context(), tenantUUID, name)
		if err != nil {
			cli.Logger.Fatalf("failed to get webhook: %v", err)
		}
		if resp.JSON200 == nil {
			cli.Logger.Fatalf("webhook %q not found (status %d)", name, resp.StatusCode())
		}

		printJSON(resp.JSON200)
	},
}

var webhooksCreateCmd = &cobra.Command{
	Use:   "create",
	Short: "Create a webhook",
	Long:  `Create an incoming webhook. All configuration is provided via flags.`,
	Example: `  # Basic auth webhook
  hatchet webhooks create --name my-hook --source GENERIC --event-key-expression "input.event" \
    --auth-type basic --username user --password pass

  # API key webhook
  hatchet webhooks create --name my-hook --source STRIPE --event-key-expression "input.type" \
    --auth-type api-key --api-key-header X-Api-Key --api-key secret

  # HMAC webhook
  hatchet webhooks create --name my-hook --source GITHUB --event-key-expression "input.action" \
    --auth-type hmac --hmac-signing-secret secret --hmac-signature-header X-Hub-Signature-256 \
    --hmac-algorithm SHA256 --hmac-encoding HEX`,
	Run: func(cmd *cobra.Command, args []string) {
		isJSON := isJSONOutput(cmd)
		_, hatchetClient := clientFromCmd(cmd)
		tenantUUID := clientTenantUUID(hatchetClient)

		name, _ := cmd.Flags().GetString("name")
		sourceStr, _ := cmd.Flags().GetString("source")
		eventKeyExpression, _ := cmd.Flags().GetString("event-key-expression")
		authTypeStr, _ := cmd.Flags().GetString("auth-type")

		if name == "" {
			cli.Logger.Fatal("--name is required")
		}
		if eventKeyExpression == "" {
			cli.Logger.Fatal("--event-key-expression is required")
		}

		source, err := parseWebhookSourceName(sourceStr)
		if err != nil {
			cli.Logger.Fatalf("%v", err)
		}

		var body rest.V1CreateWebhookRequest

		switch strings.ToLower(strings.ReplaceAll(strings.TrimSpace(authTypeStr), "_", "-")) {
		case "basic":
			username, _ := cmd.Flags().GetString("username")
			password, _ := cmd.Flags().GetString("password")
			if username == "" || password == "" {
				cli.Logger.Fatal("--username and --password are required for basic auth")
			}
			err = body.FromV1CreateWebhookRequestBasicAuth(rest.V1CreateWebhookRequestBasicAuth{
				AuthType:           rest.BASIC,
				Auth:               rest.V1WebhookBasicAuth{Username: username, Password: password},
				Name:               name,
				SourceName:         source,
				EventKeyExpression: eventKeyExpression,
			})
		case "api-key", "apikey":
			apiKeyHeader, _ := cmd.Flags().GetString("api-key-header")
			apiKey, _ := cmd.Flags().GetString("api-key")
			if apiKeyHeader == "" || apiKey == "" {
				cli.Logger.Fatal("--api-key-header and --api-key are required for api-key auth")
			}
			err = body.FromV1CreateWebhookRequestAPIKey(rest.V1CreateWebhookRequestAPIKey{
				AuthType:           rest.V1CreateWebhookRequestAPIKeyAuthTypeAPIKEY,
				Auth:               rest.V1WebhookAPIKeyAuth{HeaderName: apiKeyHeader, ApiKey: apiKey},
				Name:               name,
				SourceName:         source,
				EventKeyExpression: eventKeyExpression,
			})
		case "hmac":
			signingSecret, _ := cmd.Flags().GetString("hmac-signing-secret")
			signatureHeader, _ := cmd.Flags().GetString("hmac-signature-header")
			algorithmStr, _ := cmd.Flags().GetString("hmac-algorithm")
			encodingStr, _ := cmd.Flags().GetString("hmac-encoding")
			if signingSecret == "" || signatureHeader == "" {
				cli.Logger.Fatal("--hmac-signing-secret and --hmac-signature-header are required for hmac auth")
			}
			algorithm, algErr := parseWebhookHMACAlgorithm(algorithmStr)
			if algErr != nil {
				cli.Logger.Fatalf("%v", algErr)
			}
			encoding, encErr := parseWebhookHMACEncoding(encodingStr)
			if encErr != nil {
				cli.Logger.Fatalf("%v", encErr)
			}
			err = body.FromV1CreateWebhookRequestHMAC(rest.V1CreateWebhookRequestHMAC{
				AuthType: rest.HMAC,
				Auth: rest.V1WebhookHMACAuth{
					SigningSecret:       signingSecret,
					SignatureHeaderName: signatureHeader,
					Algorithm:           algorithm,
					Encoding:            encoding,
				},
				Name:               name,
				SourceName:         source,
				EventKeyExpression: eventKeyExpression,
			})
		default:
			cli.Logger.Fatalf("invalid auth type %q (must be one of: basic, api-key, hmac)", authTypeStr)
		}

		if err != nil {
			cli.Logger.Fatalf("failed to build webhook request: %v", err)
		}

		resp, err := hatchetClient.API().V1WebhookCreateWithResponse(cmd.Context(), tenantUUID, body)
		if err != nil {
			cli.Logger.Fatalf("failed to create webhook: %v", err)
		}
		if resp.JSON200 == nil {
			cli.Logger.Fatalf("failed to create webhook: %s (status %d)", apiErrorMessage(resp.JSON400, resp.JSON403, resp.JSON404), resp.StatusCode())
		}

		if !isJSON {
			fmt.Println(styles.SuccessMessage(fmt.Sprintf("Created webhook: %s", resp.JSON200.Name)))
		}
		printJSON(resp.JSON200)
	},
}

var webhooksUpdateCmd = &cobra.Command{
	Use:   "update <webhook-name>",
	Short: "Update a webhook",
	Long:  `Update a webhook by name. Only the provided flags are changed.`,
	Args:  cobra.ExactArgs(1),
	Example: `  hatchet webhooks update my-hook --event-key-expression "input.event_type"
  hatchet webhooks update my-hook --scope-expression "input.tenant" --return-event-as-response-payload`,
	Run: func(cmd *cobra.Command, args []string) {
		name := args[0]
		isJSON := isJSONOutput(cmd)
		_, hatchetClient := clientFromCmd(cmd)
		tenantUUID := clientTenantUUID(hatchetClient)

		body := rest.V1UpdateWebhookRequest{}
		changed := false

		if cmd.Flags().Changed("event-key-expression") {
			v, _ := cmd.Flags().GetString("event-key-expression")
			body.EventKeyExpression = &v
			changed = true
		}
		if cmd.Flags().Changed("scope-expression") {
			v, _ := cmd.Flags().GetString("scope-expression")
			body.ScopeExpression = &v
			changed = true
		}
		if cmd.Flags().Changed("return-event-as-response-payload") {
			v, _ := cmd.Flags().GetBool("return-event-as-response-payload")
			body.ReturnEventAsResponsePayload = &v
			changed = true
		}
		if cmd.Flags().Changed("static-payload") {
			v, _ := cmd.Flags().GetString("static-payload")
			payload, err := parseJSONObjectFlag("static-payload", v)
			if err != nil {
				cli.Logger.Fatalf("%v", err)
			}
			body.StaticPayload = &payload
			changed = true
		}

		if !changed {
			cli.Logger.Fatal("no fields to update; provide at least one flag")
		}

		resp, err := hatchetClient.API().V1WebhookUpdateWithResponse(cmd.Context(), tenantUUID, name, body)
		if err != nil {
			cli.Logger.Fatalf("failed to update webhook: %v", err)
		}
		if resp.JSON200 == nil {
			cli.Logger.Fatalf("failed to update webhook: %s (status %d)", apiErrorMessage(resp.JSON400, resp.JSON403, resp.JSON404), resp.StatusCode())
		}

		if isJSON {
			printJSON(resp.JSON200)
		} else {
			fmt.Println(styles.SuccessMessage(fmt.Sprintf("Updated webhook: %s", resp.JSON200.Name)))
		}
	},
}

var webhooksDeleteCmd = &cobra.Command{
	Use:   "delete <webhook-name>",
	Short: "Delete a webhook",
	Long:  `Delete a webhook by name. Use --yes to skip confirmation.`,
	Args:  cobra.ExactArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		name := args[0]
		isJSON := isJSONOutput(cmd)
		yes, _ := cmd.Flags().GetBool("yes")
		_, hatchetClient := clientFromCmd(cmd)
		tenantUUID := clientTenantUUID(hatchetClient)

		if !isJSON && !yes {
			if !confirmAction(fmt.Sprintf("Delete webhook '%s'?", name)) {
				fmt.Println("Aborted.")
				return
			}
		}

		resp, err := hatchetClient.API().V1WebhookDeleteWithResponse(cmd.Context(), tenantUUID, name)
		if err != nil {
			cli.Logger.Fatalf("failed to delete webhook: %v", err)
		}
		if resp.StatusCode() >= 400 {
			cli.Logger.Fatalf("failed to delete webhook: %s (status %d)", apiErrorMessage(resp.JSON400, resp.JSON403, resp.JSON404), resp.StatusCode())
		}

		if isJSON {
			printJSON(map[string]interface{}{"deleted": true, "name": name})
		} else {
			fmt.Println(styles.SuccessMessage(fmt.Sprintf("Deleted webhook: %s", name)))
		}
	},
}

func parseWebhookSourceName(s string) (rest.V1WebhookSourceName, error) {
	v := rest.V1WebhookSourceName(strings.ToUpper(strings.TrimSpace(s)))
	for _, valid := range webhookSourceNames {
		if v == valid {
			return v, nil
		}
	}
	names := make([]string, len(webhookSourceNames))
	for i, valid := range webhookSourceNames {
		names[i] = string(valid)
	}
	return "", fmt.Errorf("invalid source %q (must be one of: %s)", s, strings.Join(names, ", "))
}

func parseWebhookHMACAlgorithm(s string) (rest.V1WebhookHMACAlgorithm, error) {
	v := rest.V1WebhookHMACAlgorithm(strings.ToUpper(strings.TrimSpace(s)))
	for _, valid := range webhookHMACAlgorithms {
		if v == valid {
			return v, nil
		}
	}
	names := make([]string, len(webhookHMACAlgorithms))
	for i, valid := range webhookHMACAlgorithms {
		names[i] = string(valid)
	}
	return "", fmt.Errorf("invalid hmac algorithm %q (must be one of: %s)", s, strings.Join(names, ", "))
}

func parseWebhookHMACEncoding(s string) (rest.V1WebhookHMACEncoding, error) {
	v := rest.V1WebhookHMACEncoding(strings.ToUpper(strings.TrimSpace(s)))
	for _, valid := range webhookHMACEncodings {
		if v == valid {
			return v, nil
		}
	}
	names := make([]string, len(webhookHMACEncodings))
	for i, valid := range webhookHMACEncodings {
		names[i] = string(valid)
	}
	return "", fmt.Errorf("invalid hmac encoding %q (must be one of: %s)", s, strings.Join(names, ", "))
}

func apiErrorMessage(errs ...*rest.APIErrors) string {
	for _, e := range errs {
		if e == nil {
			continue
		}
		msgs := make([]string, 0, len(e.Errors))
		for _, apiErr := range e.Errors {
			msgs = append(msgs, apiErr.Description)
		}
		if len(msgs) > 0 {
			return strings.Join(msgs, "; ")
		}
	}
	return "unexpected response from API"
}

func init() {
	rootCmd.AddCommand(webhooksCmd)
	webhooksCmd.AddCommand(webhooksListCmd, webhooksGetCmd, webhooksCreateCmd, webhooksUpdateCmd, webhooksDeleteCmd)

	webhooksCmd.PersistentFlags().StringP("profile", "p", "", "Profile to use for connecting to Hatchet (default: prompts for selection)")
	webhooksCmd.PersistentFlags().StringP("output", "o", "", "Output format: json (skips interactive TUI)")

	webhooksCreateCmd.Flags().StringP("name", "n", "", "Name of the webhook")
	webhooksCreateCmd.Flags().StringP("source", "s", string(rest.GENERIC), "Webhook source: GENERIC, GITHUB, LINEAR, SLACK, STRIPE, SVIX")
	webhooksCreateCmd.Flags().String("event-key-expression", "", "CEL expression used to create the event key from the webhook payload")
	webhooksCreateCmd.Flags().String("auth-type", "", "Authentication type: basic, api-key, hmac")
	webhooksCreateCmd.Flags().String("username", "", "Username for basic auth")
	webhooksCreateCmd.Flags().String("password", "", "Password for basic auth")
	webhooksCreateCmd.Flags().String("api-key-header", "", "Header name for api-key auth")
	webhooksCreateCmd.Flags().String("api-key", "", "API key for api-key auth")
	webhooksCreateCmd.Flags().String("hmac-signing-secret", "", "Signing secret for hmac auth")
	webhooksCreateCmd.Flags().String("hmac-signature-header", "", "Signature header name for hmac auth")
	webhooksCreateCmd.Flags().String("hmac-algorithm", string(rest.SHA256), "HMAC algorithm: MD5, SHA1, SHA256, SHA512")
	webhooksCreateCmd.Flags().String("hmac-encoding", string(rest.HEX), "HMAC encoding: BASE64, BASE64URL, HEX")

	webhooksUpdateCmd.Flags().String("event-key-expression", "", "CEL expression used to create the event key from the webhook payload")
	webhooksUpdateCmd.Flags().String("scope-expression", "", "CEL expression used to filter the correct workflow to trigger")
	webhooksUpdateCmd.Flags().Bool("return-event-as-response-payload", false, "Return the triggered event as the response payload")
	webhooksUpdateCmd.Flags().String("static-payload", "", "Static payload to send with the webhook, as a JSON object")

	webhooksDeleteCmd.Flags().BoolP("yes", "y", false, "Skip confirmation prompt")
}
